lib/core/utils/selector-cache.js

```diff
@@ -0,0 +1,196 @@
+import { matchesExpression } from './matches';
+import tokenList from './token-list';
+
+// since attribute names can't contain whitespace, this will be
+// a reserved list for ids so we can perform virtual id lookups
+const idsKey = ' [idsMap]';
+let selectorMap = {};
+
+/**
+ * Non-tag selectors use `*` for the tag name so a global selector won't have any other properties of the expression.
+ * @param {Object} exp Selector Expression
+ * @returns {Boolean}
+ */
+function isGlobalSelector(exp) {
+  return exp.tag === '*' && !exp.attributes && !exp.id && !exp.classes;
```

Comment:
Since we don't cache the content of a pseudo selector, they are still considered a global selector such that `*:not([foo])` will use the global selector cache and then `matchesExpression` to match the pseudo selector.